monitor: multiple readers support

aalemayhu opened the pull request:
Please see individual patches for the details.

tgraf commented on pkg/bpf/monitor_payload.go:
```diff
@@ -0,0 +1,63 @@
+// MonitorPayloadMeta is used by readers to get information about the payload.
+type MonitorPayloadMeta struct {
+	Size uint32
+	_    [28]byte // Reversed 28 bytes for future fields.
```
Reserved

tgraf commented on cilium/cmd/monitor.go:
```diff
@@ -235,35 +234,50 @@ func runMonitor() {
-		events.Cpus, events.Npages, events.Pagesize)
 	fmt.Printf("Press Ctrl-C to quit\n")
 
+	betaBuf := make([]byte, 32)
```
Can we use `unsafe.Sizeof(bpf.MonitorPayloadMeta)` here?

tgraf commented on daemon/node_monitor.go:
```diff
@@ -0,0 +1,151 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
```
Does it make sense to move this to `pkg/monitor` instead? We we want to move away from `daemon/` in the future.

aalemayhu replied:
Yes, will also move the payload stuff there.

tgraf commented on daemon/node_monitor.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// EnableMonitor is used by the daemon to turn on the traffic monitoring.
+func (d *Daemon) EnableMonitor() {
```
As you note correctly. The big downside of this is that any crash of the monitor brings down the entire agent. I think we should `os.Exec` the handler here and restart it when it crashes. That we don't risk crashing the entire agent. What do you think?

aalemayhu replied:
So the exec package is for running external commands. Do you want this to be a separate binary that is executed?

If that's not the case then I would continue with moving this to the  monitor package and look into keeping a reference to a new structure `NodeMonitor` to remove most of the daemon references. The daemon would configure this struct and instead of using fatal, return errors  and handle them.

aalemayhu replied:
Please ignore the previous comment above. It was written before today's call. Will continue  moving this into a separate binary. Thanks.

tgraf reviewed (changes requested):
This is great work! Looking very good.

aalemayhu committed:
monitor: new package for running multiple monitors

The monitor directory consists of two convenience packages, but mainly
it produces a binary(`cilium-node-monitor`). The node monitor starts
listening to the perf ring buffer and accepts reader connections via a
UDS from the runtime directory. The first argument to the node monitor
affects the number of pages used. Every 5 seconds the monitor state is
printed out as json to stdout.

The payload used by the readers is in `./monitor/payload`. The other
package is `./monitor/launch` and it is supposed to make it easier to
rerun the binary programmatically. For example in the scenario where you
would like to modify the configuration.

Related-to: #1072 (Allow multiple listeners to cilium monitor)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
tests: add test for running multiple monitors

Comparing the output from 3 monitors.

Closes: #1072 (Allow multiple listeners to cilium monitor)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
api: add models for keeping track of the node monitor

`MonitorStatus` contains all the relevant monitor fields. This will be
used both for showing the status and reading / modifying the
configuration. `Npages` is the only field which will be modified for
now.

Related-to: #1072 (Allow multiple listeners to cilium monitor)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
cilium/cmd: use monitor package for reading events

The daemon is used for getting the monitor state and the
configuration options are available via `cilium config`.

Also fixed one typo.

Related-to: #1072 (Allow multiple listeners to cilium monitor)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
daemon/defaults: add monitor UDS path

will be used in follow up commit by the monitor code.

Related-to: #1072 (Allow multiple listeners to cilium monitor)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
daemon: configure and use a instance of the NodeMonitor

. `cilium status` shows the status summary of the node monitor.
. `cilium config` shows and can modify the number of pages.

Related-to: #1072 (Allow multiple listeners to cilium monitor)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

tgraf commented on pkg/monitor/launch/launcher.go:
```diff
@@ -0,0 +1,89 @@
+
+const targetName = "cilium-node-monitor"
+
+var stateMutex sync.Mutex
```
Please add a small comment which documents what this lock protects

aalemayhu replied:
Will drop it. Added it because of some other conditional issue, don't think it's needed.

tgraf commented on pkg/monitor/launch/launcher.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// StateLocked returns the monitor status.
+func (nm *NodeMonitor) StateLocked() *models.MonitorStatus {
```
We typically use the `Locked()` suffix to indicate that the structure must already be locked.

tgraf commented on daemon/daemon.go:
```diff
@@ -1086,6 +1089,13 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 
+	if numPagesEntry, ok := params.Configuration["MonitorNumPages"]; ok {
+		if d.NodeMonitor.Arg != numPagesEntry {
+			d.NodeMonitor.Restart(numPagesEntry)
```
Awesome! 👍

tgraf reviewed (changes requested):
This looks fantastic. Great work @scanf!

aanm commented on api/v1/server/embedded_spec.go:
```diff
@@ -1262,6 +1266,26 @@ func init() {
+        "Cpus": {
+          "type": "integer"
+        },
+        "Lost": {
```
Does swagger allow to add a description to each element? I'm kind of lost what `Lost` means.

aalemayhu replied:
Added.

aanm commented on api/v1/server/restapi/cilium_api.go:
```diff
@@ -401,6 +401,9 @@ func (o *CiliumAPI) HandlerFor(method, path string) (http.Handler, bool) {
 		return nil, false
 	}
+	if path == "/" {
+		path = ""
```
Who wrote this? Was automatic?

aalemayhu replied:
all the files in the api/v1 directory were modified using the generate-api target. This is probably part of the template go-swagger uses https://github.com/go-swagger/go-swagger/blob/master/generator/templates/server/builder.gotmpl#L268-L270

aanm commented on pkg/monitor/launch/launcher.go:
```diff
@@ -0,0 +1,78 @@
+			log.Errorf("cmd.Start(): %s", err)
+		}
+		nm.process = cmd.Process
+		go func() { cmd.Wait() }()
```
What is this line for? A comment would help.

aalemayhu replied:
This is a leftover from when I was trying to get `ProcessState` from the command but it was set after `Wait` is called. Thanks!

aanm commented on pkg/monitor/launch/launcher.go:
```diff
@@ -0,0 +1,78 @@
+
+		r := bufio.NewReader(stdout)
+		for nm.process != nil {
+			l, _ := r.ReadString('\n')
```
Since every line of code matters to the monitor, did you find any performance difference between `ReadString` and `ReadBytes` or `NewReader` and `NewScanner`?

aalemayhu replied:
Great point. Have not spent time on measuring the difference between those.
Getting the monitor processes output of the top top output while running a netperf
benchmark  would be nice, but if you have other ideas on how to measure difference, please share.

aanm replied:
I was thinking with this https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

aalemayhu replied:
Will look into the post, might take sometime to create a usable benchmark. Thanks!

aanm commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,175 @@
+	// Dump stat
+	t := time.NewTicker(5 * time.Second)
+	go func() {
+		for {
```
```
go func(t time.Ticker)
for _ := range <-t.C {
    if monitorEvents != nil {
	dumpStat()
  }
}(t)
```

aalemayhu replied:
Passing the parameter t is fine, but not sure about the for loop there.

aanm replied:
The `for { select{ case:` that you wrote is doing the same thing that `for _ := range <- t.C` is doing

aalemayhu replied:
Tried your suggestion but just seeing compile error, what am I missing?

```shell
$ make
go build -i -ldflags '-X "github.com/cilium/cilium/pkg/version.Version=0.11.90 7dcf188f028d Wed, 23 Aug 2017 11:45:29 +0200 go version devel +e9348ab4e9c8 Fri Aug 11 05:08:44 2017 +0000 linux/amd64"' -o cilium-node-monitor
# github.com/cilium/cilium/pkg/monitor
./main.go:103:12: no new variables on left side of :=
./main.go:103:12: cannot range over <-t.C (type time.Time)
make: *** [Makefile:6: cilium-node-monitor] Error 2
```

```diff
diff --git a/pkg/monitor/main.go b/pkg/monitor/main.go
index a023db5e9e01..dfa15c20b0cd 100644
--- a/pkg/monitor/main.go
+++ b/pkg/monitor/main.go
@@ -100,12 +100,9 @@ func runMonitor(npages int) {
        // Dump stat
        t := time.NewTicker(5 * time.Second)
        go func(t *time.Ticker) {
-               for {
-                       select {
-                       case <-t.C:
-                               if monitorEvents != nil {
-                                       dumpStat()
-                               }
+               for _ := range <-t.C {
+                       if monitorEvents != nil {
+                               dumpStat()
                        }
                }
        }(t)
```

aanm commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,175 @@
+				}
+			}
+		}
+	}()
```
If the main event loop on line 114 exits, is the main program closed? If No then add
`defer t.Stop()`

aalemayhu replied:
Yes, but the program can also be killed by the daemon on configuration changes.

aanm commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,175 @@
+	buf := pl.Encode()
+	meta := &payload.Meta{Size: uint32(len(buf))}
+	i := 0 // Used for removing lost clients.
+	for _, client := range listeners {
```
Does the order of the clients matter? All of them will receive the message sooner or later. On that case you can implement this with a map instead of a slice and you don't need do a new assignment on line 150, you can remove the `i` variable and...

aalemayhu replied:
No order should not matter. Map looks much nicer, thanks.

aanm commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,175 @@
+		if err := binary.Write(client, byteorder.Native, meta); err != nil {
+			log.Infof("Removing lost client %s", err)
+			client.Close()
+			continue
```
... here you can do `delete(listeners, client)`

aalemayhu replied:
Done.

aanm commented on pkg/monitor/payload/monitor_payload.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+const (
+	// EventSample is equivalent to PERF_RECORD_SAMPLE
```
`is equivalent to PERF_RECORD_SAMPLE` it would help to know from which file is this on

aalemayhu replied:
Done.

aanm commented on pkg/monitor/payload/monitor_payload.go:
```diff
@@ -0,0 +1,63 @@
+// Meta is used by readers to get information about the payload.
+type Meta struct {
+	Size uint32
+	_    [28]byte // Reserved 28 bytes for future fields.
```
Why `28`?

aalemayhu replied:
I was asking @tgraf for feedback on Slack and he wanted to use 32 bytes instead of just a 4 byte length. I don't know what those 28 remaining bytes will be used for in the future.

aanm commented on cilium/cmd/monitor.go:
```diff
@@ -235,35 +240,46 @@ func runMonitor() {
+		// how much data to read.
+		if _, err := conn.Read(metaBuf); err != nil {
+			if err == io.EOF {
+				time.Sleep(12 * time.Second)
```
Why 12 seconds?

aalemayhu replied:
In the reload make target, it takes normally 6 seconds. Observed that it sometimes takes more time for the daemon to restart. So doubled it.

ashwinp replied:
Please define a const in this file so that it will be easier to update it.

aanm commented on cilium/cmd/monitor.go:
```diff
@@ -235,35 +240,46 @@ func runMonitor() {
 	signal.Notify(signalChan, os.Interrupt)
 	go func() {
```
Because of the jump you're doing on line 226, you're are also initializing this go routine

aalemayhu replied:
Moved it out. Thanks.

aanm commented on tests/17-multiple-monitors.sh:
```diff
@@ -0,0 +1,53 @@
+kill ${MON_1_PID} ${MON_2_PID} ${MON_3_PID}
+
+if ! diff3 ${MON_1_OUTPUT} ${MON_2_OUTPUT} ${MON_3_OUTPUT}; then
+  #vimdiff ${MON_1_OUTPUT} ${MON_2_OUTPUT} ${MON_3_OUTPUT}
```
Left over?

aalemayhu replied:
Actually I was considering leaving this in because it was so useful, but will remove. Thanks.

aanm reviewed (commented):
If I understood correctly, is the monitor is always running?
Where is the line where you add a new listener / client to that monitor?

What was the motivation to have a new binary for the monitor?

rlenglet commented on pkg/monitor/Makefile:
```diff
@@ -0,0 +1,16 @@
+include ../../Makefile.defs
```
Makefiles should have a copyright statement like any source code file.

aalemayhu replied:
Fixed

rlenglet commented on api/v1/server/embedded_spec.go:
```diff
@@ -873,6 +873,10 @@ func init() {
       "description": "Response to a daemon configuration request. Contains the addressing\ninformation and configuration settings.\n",
       "type": "object",
       "properties": {
+        "Monitor": {
```
Capitalization is not consistent. All properties start in lowercase.
Also, this is not consistent with the new "nodeMonitor" property in StatusResponse, which has the same type.
Use the "nodeMonitor" too here for consistency.

aalemayhu replied:
Fixed

rlenglet commented on api/v1/server/embedded_spec.go:
```diff
@@ -1262,6 +1266,31 @@ func init() {
+    "MonitorStatus": {
+      "description": "Status of the node monitor",
+      "properties": {
+        "Cpus": {
```
Property names should start in lowercase.

aalemayhu replied:
Fixed

rlenglet commented on api/v1/server/embedded_spec.go:
```diff
@@ -1262,6 +1266,31 @@ func init() {
+      "description": "Status of the node monitor",
+      "properties": {
+        "Cpus": {
+          "description": "The number of CPUs to listen on for events.\n",
```
The style in this Swagger spec is to avoid "The".
Also, remove the trailing newline.

    "description": "Number of CPUs to listen on for events.",

aalemayhu replied:
Fixed

rlenglet commented on api/v1/models/monitor_status.go:
```diff
@@ -0,0 +1,45 @@
+
+	// The number of CPUs to listen on for events.
+	//
+	Cpus int64 `json:"Cpus,omitempty"`
```
Fix the capitalization of the JSON property name, cf. my comments on embedded_spec.go.

rlenglet replied:
Same for the other fields below.

rlenglet commented on api/v1/models/monitor_status.go:
```diff
@@ -0,0 +1,45 @@
+// swagger:model MonitorStatus
+type MonitorStatus struct {
+
+	// The number of CPUs to listen on for events.
```
Fix the godoc comment, by starting with the name of the field:

    // Cpus is the number ...

rlenglet replied:
Same for the other comments below.

rlenglet commented on api/v1/models/monitor_status.go:
```diff
@@ -0,0 +1,45 @@
+// Editing this file might prove futile when you re-run the swagger generate command
+
+import (
+	strfmt "github.com/go-openapi/strfmt"
```
Remove the redundant package name:

    "github.com/go-openapi/strfmt"

rlenglet commented on api/v1/models/daemon_configuration_response.go:
```diff
@@ -16,6 +16,9 @@ import (
 type DaemonConfigurationResponse struct {
 
+	// Status of the node monitor
+	Monitor *MonitorStatus `json:"Monitor,omitempty"`
```
Monitor -> nodeMonitor, for consistency.

rlenglet commented on api/v1/models/daemon_configuration_response.go:
```diff
@@ -49,6 +57,25 @@ func (m *DaemonConfigurationResponse) Validate(formats strfmt.Registry) error {
+
+		if err := m.Monitor.Validate(formats); err != nil {
+			if ve, ok := err.(*errors.Validation); ok {
+				return ve.ValidateName("Monitor")
```
"Monitor" -> "nodeMonitor" for consistency.

rlenglet commented on api/v1/openapi.yaml:
```diff
@@ -939,6 +966,9 @@ definitions:
         type: string
       k8s-configuration:
         type: string
+      Monitor:
```
Monitor -> nodeMonitor, for consistency.

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
```
Move this particular file into /monitor instead of /pkg/monitor. There are currently no main packages in /pkg. All main packages are in top-level directories: /cilium, /daemon, /plugins.
Update /Makefile accordingly.

aalemayhu replied:
Fixed

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+	for {
+		conn, err := server.Accept()
+		if err != nil {
+			log.Print(err)
```
Log at warning level:

    log.WithError(err).Warn("error accepting connection")

aalemayhu replied:
Fixed

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+var (
+	listeners     = make(map[int]net.Conn)
```
A map seems inefficient here. I would use a different data structure. You seem to need a data structure that makes it cheap to add elements, allows iterating, and easily removing elements while iterating. I would recommend using a list:

    listeners = list.New()

All your operations would be O(1).

aalemayhu replied:
cool, did not know about it.

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+		}
+
+		mutex.Lock()
+		listeners[len(listeners)+1] = conn
```
If using a list:

    listeners.PushBack(conn)

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+
+		mutex.Lock()
+		listeners[len(listeners)+1] = conn
+		log.Infof("New monitor connected. len(listeners) = %d", len(listeners))
```
If using a list:
len(listeners) -> listeners.Len()

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+
+	mp, err := json.Marshal(m)
+	if err != nil {
+		log.Errorf("json: %s", err)
```
log.WithError(err).Error("error marshalling JSON")

aalemayhu replied:
Fixed

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+func send(pl payload.Payload) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	if len(listeners) == 0 {
```
If using a list:

    if listeners.Len() == 0 {

rlenglet commented on pkg/monitor/main.go:
```diff
@@ -0,0 +1,172 @@
+
+	buf := pl.Encode()
+	meta := &payload.Meta{Size: uint32(len(buf))}
+	for k, client := range listeners {
```
If using a list:

    for e := listeners.Front(); e != nil; e = e.Next() {
        client := e.Value().(net.Conn)
        ...
        // To delete:
        listeners.Remove(e)
    }

rlenglet replied:
Just have to be careful that Remove(e) sets e.Next() to nil. Use another variable to hold the next element before removing:

    for e := listeners.Front(); e != nil; {
        next := e.Next()
        client := e.Value().(net.Conn)
        ...
        // To delete:
        listeners.Remove(e)
        ...
        e = next
    }

aalemayhu replied:
Not sure I understand. Why isn't the first loop good enough?

rlenglet commented on cilium/cmd/config.go:
```diff
@@ -67,21 +71,25 @@ func dumpConfig(Opts map[string]string) {
+	}
+
+	if numPages > 0 && numPages != int(resp.NodeMonitor.Npages) {
+		fmt.Printf("%d -> %d\n", int(resp.NodeMonitor.Npages), numPages)
```
Remove? I suppose this was for debugging?

aalemayhu replied:
Fixed

rlenglet commented on daemon/daemon.go:
```diff
@@ -118,6 +119,8 @@ type Daemon struct {
 	uniqueIDMU sync.Mutex
 	uniqueID   map[uint64]bool
+
+	NodeMonitor *monitor.NodeMonitor
```
Why export this field, unlike the others?
NodeMonitor -> nodeMonitor

aalemayhu replied:
Fixed

rlenglet commented on cilium/cmd/config.go:
```diff
@@ -46,6 +49,7 @@ var configCmd = &cobra.Command{
 func init() {
 	RootCmd.AddCommand(configCmd)
 	configCmd.Flags().BoolVarP(&listOptions, "list-options", "", false, "List available options")
+	configCmd.Flags().IntVarP(&numPages, "num-pages", "n", 0, "Number of pages for perf ring buffer")
```
The "0" default has a special meaning. Document it in the flag's doc string, e.g. "Number of pages [...]; if 0, defaults to the Cilium daemon's configuration"

aalemayhu replied:
Fixed

rlenglet commented on monitor/main.go:
```diff
@@ -0,0 +1,174 @@
+		if err := binary.Write(client, byteorder.Native, meta); err != nil {
+			log.Infof("Removing lost client %s", err)
+			client.Close()
+			listeners.Remove(e)
```
Beware that Remove(e) sets e.Next to nil, so you have to preserve e.Next() before removing e.

aalemayhu replied:
please see the latest diff.

aalemayhu committed:
docker: add the monitor to the list of directories to be copied

Fixes a build error where the docker files are missing the monitor
package. Also added a comment above the ADD keyword to clarify where the
list of directory entries to be added are.

Related-to: #1072 (Allow multiple listeners to cilium monitor)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

ashwinp commented on api/v1/openapi.yaml:
```diff
@@ -942,6 +946,24 @@ definitions:
+        description: Pages size used for the perf ring buffer.
+        type: integer
+      lost:
+        description: Recorded sample number lost by perf.
```
Is this the number of samples lost/dropped by perf? If yes, please update this to "Number of samples lost by perf."

aalemayhu replied:
Yes, the number of lost events comes from one of the perf structures.

ashwinp commented on api/v1/openapi.yaml:
```diff
@@ -942,6 +946,24 @@ definitions:
+        description: Recorded sample number lost by perf.
+        type: integer
+      unknown:
+        description: Number of unknown samples to cilium.
```
you can remove "to cilium"

aalemayhu replied:
Fixed

ashwinp commented on cilium/cmd/monitor.go:
```diff
@@ -201,7 +202,7 @@ func receiveEvent(msg *bpf.PerfEventSample, cpu int) {
 		captureEvents(prefix, data)
 	default:
-		fmt.Printf("%s Unknonwn event: %+v\n", prefix, msg)
+		fmt.Printf("%s Unknonwn event: %+v\n", prefix, data)
```
typo: Unknown

aalemayhu replied:
That was already there but fixed. Thanks!

ashwinp commented on cilium/cmd/monitor.go:
```diff
@@ -218,52 +219,70 @@ func validateEventTypeFilter() {
+		if _, err := conn.Read(metaBuf); err != nil {
+			if err == io.EOF {
+				time.Sleep(12 * time.Second)
+				//conn.Close()
```
remove commented code

aalemayhu replied:
Fixed

ashwinp commented on monitor/Makefile:
```diff
@@ -0,0 +1,30 @@
+# Copyright 2016- 2017Authors of Cilium
```
space after 2017

aalemayhu replied:
Fixed

ashwinp commented on monitor/main.go:
```diff
@@ -0,0 +1,176 @@
+	monitorEvents *bpf.PerCpuEvents
+)
+
+// handleMonitorConn handles all the incoming monitors.
```
incoming monitor connections?

aalemayhu replied:
Fixed

ashwinp commented on monitor/main.go:
```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+func send(pl payload.Payload) {
```
GoDoc is missing

tgraf replied:
This is a private function

aalemayhu replied:
Fixed

ashwinp commented on monitor/main.go:
```diff
@@ -0,0 +1,176 @@
+	monitorEvents *bpf.PerCpuEvents
+)
+
+// handleMonitorConn handles all the incoming monitors.
```
run, dumpStat, handleConnection, send can be defined as methods of a new type Monitor. You could also move the vars in lines 51 to 54 as instance variables of the type.

aalemayhu replied:
Fixed

ashwinp commented on monitor/main.go:
```diff
@@ -0,0 +1,176 @@
+
+	// Main event loop
+	for {
+		todo, err := monitorEvents.Poll(5000)
```
Please define a const for 5000.

tgraf replied:
This is existing code

aalemayhu replied:
Fixed

ashwinp commented on monitor/main.go:
```diff
@@ -0,0 +1,176 @@
+}
+
+func main() {
+	common.RequireRootPrivilege(targetName)
```
Please add doc to explain why we need root privileges.

tgraf replied:
This is existing code

aalemayhu replied:
Fixed

ashwinp commented on tests/17-multiple-monitors.sh:
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env bash
+
+source ./helpers.bash
+
```
Please add `set -uex` to the script.

aalemayhu replied:
Fixed

ashwinp reviewed (approved):
LGTM. Added minor nitpicks/comments.
